### PR TITLE
Team4api

### PR DESCRIPTION
**Pull request by hepengcs**
Inventory Management APIs are completed. 
Following issues need to be solved ASAP.
1. What status code should be sent to UI when CRUD operation succeeds or fails?  We're supposed to use the same code list.
2. Json representation of validation errors should be defined.

### CODE REVIEW

File: app/Http/Controllers/Api/InventoryController1.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
+<?php namespace App\Http\Controllers;
```
Comment (zachmay) [inline app/Http/Controllers/Api/InventoryController1.php]: The namespace for this class should be `App\Http\Controllers\Api`. The Laravel framework needs the path to the class definition file to correspond with the class's namespace.

File: app/Http/Controllers/Api/InventoryController1.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+use App\Http\Requests;
+use App\Http\Controllers\Controller;
+use App\Item;
+
+use Illuminate\Support\Facades\Response;
+use Illuminate\Support\Facades\Input;
+
+class InventoryController1 extends Controller {
```
Comment (zachmay) [inline app/Http/Controllers/Api/InventoryController1.php]: It's generally not good to have classes named `Class1`, `Class2`. Why not rename `InventoryController1` to the more descriptive `InventoryCollectionController` and `InventoryController2` to `InventoryItemController`.

File: app/Http/Controllers/Api/InventoryController1.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+	/**
+	 * Display a listing of the resource.
+	 *
+	 * @param  string $query
+	 * @return json all or queried $items
+	 */
+	public function index($query)
```
Comment (zachmay) [inline app/Http/Controllers/Api/InventoryController1.php]: What we want this endpoint to do is return a list of all inventory items, filtered by some string. The user passes that string via [a query string](http://en.wikipedia.org/wiki/Query_string) parameter like `http://192.168.33.99/api/inventory?query=blah`.

Look at [this StackOverflow question](http://stackoverflow.com/questions/24744825/laravel-queries-strings) for how to get access to a request's query string parameters.

File: app/Http/Controllers/Api/InventoryController1.php [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	{
```
Comment (zachmay) [inline app/Http/Controllers/Api/InventoryController1.php]: In this section, we pull in all the items on line 20, then it looks like line 23 is trying to filter them. What we should do is something like:

```
if ( !is_null($query) )
{
    $items = Items::where(...)->get();
}
else
{
    $items = Items::all();
}
```

It's arguable whether we even want to allow them to retrieve an unfiltered list of *all* inventory items. Maybe throw a 400 (bad request) if they don't specify a query?

File: app/Http/Controllers/Api/InventoryController1.php [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+		$items = Item::all();
+
+		if(!is_null($query)) {
+			$items = $items->find($query);
+		}
+
+		if(is_null($items)) {
+			return Response::json("Not Found", 404);
```
Comment (zachmay) [inline app/Http/Controllers/Api/InventoryController1.php]: Think of this inventory endpoint as a collection. It always exists so it should never return 404. It's perfectly acceptable to return 200 and return JSON for an empty list. You probably don't even need a special check for that. Just `return $items;`.

File: app/Http/Controllers/Api/InventoryController1.php [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+	/**
+	 * Store a newly created resource in storage.
+	 *
+	 * @param  none
+	 * @return Json input $item
+	 */
+	public function store()
+	{
+		$input = Input::all();
```
Comment (zachmay) [inline app/Http/Controllers/Api/InventoryController1.php]: We want `Request::all()` to get the body of the POST request.

File: app/Http/Controllers/Api/InventoryController1.php [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+		if(is_null($input)) {
+			return Response::json("No Input", 406);
+		}
+		Item::create($input);
+		return $input;
```
Comment (zachmay) [inline app/Http/Controllers/Api/InventoryController1.php]: Make sure this returns a 201 status code.

File: app/Http/Controllers/Api/InventoryController1.php [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+	 * Store a newly created resource in storage.
+	 *
+	 * @param  none
+	 * @return Json input $item
+	 */
+	public function store()
+	{
+		$input = Input::all();
+		if(is_null($input)) {
+			return Response::json("No Input", 406);
```
Comment (zachmay) [inline app/Http/Controllers/Api/InventoryController1.php]: This is indeed where we would put validation, but we probably want to do more than test whether `$input` is null.

Also, we want to use HTTP 400 (bad request) if the input is invalid. 406 has to do with acceptable content types. We won't have to worry about that here.

File: app/Http/Controllers/Api/InventoryController1.php [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+		}
+		Item::create($input);
```
Comment (zachmay) [inline app/Http/Controllers/Api/InventoryController1.php]: Instead of directly returning `$input`, store the result of calling `Item::create(...)` and return that. That way if there are default values for some database columns, they'll be populated in the thing we send back.

File: app/Http/Controllers/Api/InventoryController2.php [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+	/**
+	 * Update the specified resource in storage.
+	 *
+	 * @param  int  $tag
+	 * @return Json updated $item
+	 */
+	public function update($tag)
+	{
+        $item=Item::where('asset_tag',$tag)->first();
```
Comment (zachmay) [inline app/Http/Controllers/Api/InventoryController2.php]: This indentation style can make code hard to read.

File: app/Http/Controllers/Api/InventoryController2.php [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+	/**
+	 * Remove the specified resource from storage.
+	 *
+	 * @param  int  $tag
+	 * @return Json removed $item
+	 */
+	public function destroy($tag)
+	{
+        $item=Item::where('asset_tag',$tag)->first();
```
Comment (zachmay) [inline app/Http/Controllers/Api/InventoryController2.php]: `firstOrFail()`

File: app/Http/Controllers/Api/InventoryController2.php [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+        if(is_null($item))
+        {
+            return  Response::json('Not Found',404);
+        }
+
+        $deleteItem=$item;
+        $item->delete();
+
+        return $deleteItem;
```
Comment (zachmay) [inline app/Http/Controllers/Api/InventoryController2.php]: It's arguable whether a DELETE request should return the thing it just deleted, but this is fine, I think.

File: app/Http/routes.php [hunk 13]
```diff
@@ -25,7 +25,17 @@
...
-});
+
+    #add by peng@team4
+    Route::get('inventory','InventoryController1@index');
+    Route::post('inventory','InventoryController1@store');
+    Route::get('inventory/{tag}','InventoryController2@show');
+    Route::put('inventory/{tag}','InventoryController2@update');
+    Route::delete('inventory/{tag}','InventoryController2@destroy');
+
+    Route::get('inventory-types','ItemTypeController@index');
```
Comment (zachmay) [inline app/Http/routes.php]: The inventory types collection should have URL `inventory/types`. Think of it as a "subdirectory" of inventory.

File: app/Http/Controllers/Api/ItemTypeController.php [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+			if(empty($itemTypes))
+			{
+				$httpContents = 'No item types available';
+			}
+			else
+			{
+				$httpContents = $itemTypes;
+			}
+		}
+
```
Comment (zachmay) [inline app/Http/Controllers/Api/ItemTypeController.php]: We really don't need to check whether the result of retrieving all `ItemType`s is null or empty.

Querying `ItemType::all()` should always return a list of item types. It might be empty, but that's OK. This is a collection, so if it's empty, that's not a failure case.

We want the client to be able to trust that what we send back is valid JSON that it can parse and use. If we return the string `'No item types available'`, we're breaking that rule and giving the client something it can't use programmatically.

File: app/Http/Controllers/Api/InventoryController1.php (hunk 2 above)
Comment (hepengcs) [inline app/Http/Controllers/Api/InventoryController1.php]: We name the controller classes for each team number to avoid conflict. We will merge them before integrate testing.

File: app/Http/Controllers/Api/InventoryController2.php (hunk 11 above)
Comment (hepengcs) [inline app/Http/Controllers/Api/InventoryController2.php]: Fail() is ok. It's a programming style when not using try-catch

File: app/Http/Controllers/Api/InventoryController2.php (hunk 10 above)
Comment (hepengcs) [inline app/Http/Controllers/Api/InventoryController2.php]: The indentation looks good in my local, but it's wried on git. I'm working on the display issue.

File: app/Http/Controllers/Api/InventoryController1.php (hunk 9 above)
Comment (wwen0) [inline app/Http/Controllers/Api/InventoryController1.php]: I made some changes according to your comments. But I am not sure if they
are correct. Please take a look.

What confuses me is that for the return value of "store", you first said to
return "201" and then you said to return the created item. I chose to
return "201".

Also, I am not sure what to validate in the "store" function, so I just
chose a few that I think may be important. Not sure if they are correct.
Please take a look. Thanks

**Comment by zachmay** [comment]
Nice work, guys!

File: app/Http/Controllers/Api/InventoryCollectionController.php [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+			$items = Item::where('asset_tag', $query)->orWhere('name', $query)
+			->orWhere('funding_source', $query)->orWhere('model', $query)
+			->orWhere('cpu', $query)->orWhere('ram', $query)->orWhere('os', $query)
+			->orWhere('administrator_flag', $query)->orWhere('teacher_flag', $query)
+			->orWhere('student_flag', $query)->orWhere('institution_flag', $query)->get();
+
+			return $items;
+
+		} else {
+			return Response::json("Bad Request", 400);
```
Comment (zachmay) [inline app/Http/Controllers/Api/InventoryCollectionController.php]: I think if `$query` is null (i.e., they didn't include the query string parameter) then return everything.

File: app/Http/Controllers/Api/InventoryCollectionController.php [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+	 */
+	public function store(Request $request)
+	{
+		$this->validate($request, ['asset_tag' => 'required', 'name'=>'required',
+		'administrator_flag' =>'required', 'teacher_flag'=>'required',
+		'institution_flag'=>'required']);
+
+		Item::create($request->all());
+
+		return Response::json("Created", 201);;
```
Comment (zachmay) [inline app/Http/Controllers/Api/InventoryCollectionController.php]: Return a JSON representation of the saved Item instead of just "created". 201 is the right code, but if there is an issue with duplicate asset tags, 400 is appropriate. You would want to catch that exception.
